src/lib/syncable_ledger/syncable_ledger.ml

```diff
@@ -427,6 +458,68 @@ end = struct
     if Hash.equal actual expected then `Success
     else `Hash_mismatch (expected, actual)
 
+  (* Merges each 2 contigous nodes, halving the size of the list *)
+  let rec merge_siblings : Hash.t list -> index -> Hash.t list =
+   fun nodes height ->
+    match nodes with
+    | [ l; r ] ->
+        [ Hash.merge ~height l r ]
+    | l :: r :: rest ->
+        Hash.merge ~height l r :: merge_siblings rest height
+    | _ ->
+        (* TODO: give some error as this shouldn't really happen *)
+        []
+
+  (* Assumes nodes to be a power of 2 and merges them into their common root *)
+  let rec merge_many : Hash.t list -> index -> Hash.t =
+   fun nodes depth ->
+    match nodes with
+    | [ single ] ->
+        single
+    | many ->
+        let half = merge_siblings many depth in
+        merge_many half (depth - 1)
+
+  let merge_many : Hash.t list -> index -> Hash.t =
+   fun nodes depth ->
+    let final_depth = depth + subtree_depth in
+    merge_many nodes final_depth
+
+  (* Adds the subtree given as the 2^k subtree leaves with the given prefix address *)
+  (* Returns next nodes to be checked *)
+  let add_subtree :
+         'a t
+      -> Addr.t
+      -> Hash.t list
+      -> [ `Good of (Addr.t * Hash.t) list
+         | `Hash_mismatch of Hash.t * Hash.t
+         | `Non_power ] =
+   fun t addr nodes ->
+    (* let prefix_depth = Addr.depth addr in *)
+    if Int.is_pow2 (List.length nodes) then
```

Comment:
This condition makes the protocol 'sender decides' instead of 'requester requests'. You should consider the original design I described: the requester sends the depth below the address that they would like, then the sender responds with a depth less than or equal to that, and the requester only accepts the message if the length is less than or equal.

Reply:
The issue with changing the request is that then V2 nodes wouldn't be able to connect to V1 nodes to sync, and no synced V2 node would come into existence to allow other V2s to sync, which wouldn't be softfork compatible.

Reply:
This is mediated at the network layer, as I had already mentioned [here](https://github.com/MinaProtocol/mina/pull/16093#discussion_r1771504587).